### PR TITLE
Option to add chaos to data representations to stress specialization code paths

### PR DESCRIPTION
**Pull request by chrisseaton**
```
% jt test fast -T--chaos-data
```

### CODE REVIEW

File: src/main/java/org/truffleruby/core/numeric/IntegerNodes.java [hunk 1]
```diff
@@ -1956,7 +1965,7 @@ protected Object upto(long from, long to, RubyProc block) {
 
         @Specialization
         protected Object upto(long from, double to, RubyProc block) {
-            return upto(from, (long) Math.ceil(to), block);
+            return upto(from, (long) Math.floor(to), block);
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/numeric/IntegerNodes.java:1968]: Is this change intentional?

File: src/main/java/org/truffleruby/debug/ChaosNode.java [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+    }
+
+    @Fallback
+    protected Object chaos(Object value) {
+        return value;
+    }
+
+    @TruffleBoundary
+    private boolean randomBoolean() {
+        return RANDOM.nextBoolean();
```
Comment (eregon) [inline src/main/java/org/truffleruby/debug/ChaosNode.java:69]: Could you rather use `ThreadLocalRandom.current().nextBoolean()`? Otherwise there will be a lot of contention.
I think should be random anyway to reveal more issues.

File: src/main/java/org/truffleruby/core/numeric/IntegerNodes.java [hunk 3]
```diff
@@ -1160,6 +1160,15 @@ protected Object leftShiftNeg(int a, long b) {
...
         }
 
+        @Specialization(guards = "b < 0")
+        protected Object leftShiftNeg(long a, long b) {
+            if (rightShiftNode == null) {
+                CompilerDirectives.transferToInterpreterAndInvalidate();
+                rightShiftNode = insert(RightShiftNodeFactory.create(null));
+            }
+            return rightShiftNode.executeRightShift(a, absoluteValue(b));
+        }
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/numeric/IntegerNodes.java:1170]: Does << need a similar fix?
A spec is probably best here to ensure we don't regress.

File: src/main/java/org/truffleruby/core/numeric/IntegerNodes.java (hunk 1 above)
Comment (eregon) [reply src/main/java/org/truffleruby/core/numeric/IntegerNodes.java:1968]: Because the `(int, double)` specialization uses floor() too, I see.

Comment (chrisseaton) [reply src/main/java/org/truffleruby/core/numeric/IntegerNodes.java:1968]: Right - spec for `#upto` was failing.

Comment (eregon) [reply src/main/java/org/truffleruby/core/numeric/IntegerNodes.java:1968]: Aside -- I wonder if `(long) Math.floor(to)` is the same as `(long) to` given `to` is a `double` (same for `int` instead of `long`).

Comment (chrisseaton) [reply src/main/java/org/truffleruby/core/numeric/IntegerNodes.java:1968]: If it is, I'll let the compiler work that out and do it for me!

**Review by eregon — APPROVED** [review]
